### PR TITLE
SSAST model

### PR DESCRIPTION
**Pull request by The-Mats**
Please review #3 first!

The goal is to integrate the SSAST Model into the pipeline:
- [x] Create a model class 
- [x] Implement get_embeddings function
- [x] Create config and experiment
- [x] Merge main into this branch
- [x] Recheck params, redo tests
- [x] Restructure code

### CODE REVIEW

**Comment by raphaelschwinger** [comment]
@XgamerTV Whats still to do here? If it is ready, could you merge main into it and remove the WIP?

File: configs/module/network/ssast.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+
+
+model_name: ssast
+model_type: vision
+torch_compile: False
+sampling_rate: 32_000
+normalize_spectrogram: True
+normalize_waveform: "instance_normalization"
+embedding_size: 768 # Needs to be adjusted if the other params are changed
```
Comment (raphaelschwinger) [inline configs/module/network/ssast.yaml]: @XgamerTV When what parameters change?

File: birdset/modules/models/ssast.py [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+            #  +2 for indexes because cls and dis token
+            pred[i] = self.gpredlayer(x[i, mask_index[i] + self.cls_token_num, :])
+            target[i] = input[i, mask_index[i], :]
+
+        # calculate the MSE loss
+        mse = torch.mean((pred - target) ** 2)
+
+        return mse
+
+    def forward(self, input_values, task='ft_avgtok', cluster=True, mask_patch=400, labels=None):
```
Comment (raphaelschwinger) [inline birdset/modules/models/ssast.py]: could you use a String Literal to get type safety?

File: birdset/modules/models/ssast.py [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+        self.num_patches = num_patches
+
+        self.proj = nn.Conv2d(in_chans, embed_dim, kernel_size=patch_size, stride=patch_size)
+
+    def forward(self, x):
+        x = self.proj(x).flatten(2).transpose(1, 2)
+        return x
+
+class ASTModel(nn.Module):
+    def __init__(self, label_dim=527,
```
Comment (raphaelschwinger) [inline birdset/modules/models/ssast.py]: @XgamerTV This `_init_` function is quite long. Could you split it up into smaller functions to improve readability?

File: birdset/modules/models/ssast.py [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+            self.p_f_dim, self.p_t_dim = self.get_shape(fstride, tstride, input_fdim, input_tdim, fshape, tshape)
+            num_patches = self.p_f_dim * self.p_t_dim
+            self.num_patches = num_patches
+            self.v.patch_embed.num_patches = num_patches
+            print('pretraining patch split stride: frequency={:d}, time={:d}'.format(fstride, tstride))
+            print('pretraining patch shape: frequency={:d}, time={:d}'.format(fshape, tshape))
+            print('pretraining patch array dimension: frequency={:d}, time={:d}'.format(self.p_f_dim, self.p_t_dim))
+            print('pretraining number of patches={:d}'.format(num_patches))
+
+            # the linear patch projection layer, use 1 channel for spectrogram rather than the original 3 channels for RGB images.
```
Comment (raphaelschwinger) [inline birdset/modules/models/ssast.py:139]: @XgamerTV Was SSAST trained on a 3 dimensional specrogram? If so, we probably have to do the same preprocessing to use the pretrained model.

File: birdset/modules/models/ssast.py [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+                 pretrain_stage=True, load_pretrained_mdl_path=None):
+
+        super(ASTModel, self).__init__()
+        assert timm.__version__ == '0.4.5', 'Please use timm == 0.4.5, the code might not be compatible with newer versions.'
+
+        # override timm input shape restriction
+        timm.models.vision_transformer.PatchEmbed = PatchEmbed
+
+        # pretrain the AST models
+        if pretrain_stage == True:
```
Comment (raphaelschwinger) [inline birdset/modules/models/ssast.py:98]: @XgamerTV Do you plan to pretrain the model?

File: configs/module/network/ssast.yaml (hunk 1 above)
Comment (raphaelschwinger) [reply configs/module/network/ssast.yaml]: When using different model sizes?

**Review by raphaelschwinger — COMMENTED** [review]
@XgamerTV Thanks for your work. Really good results!
Please add as a comment the source of where you find the code.
And please structure the code a bit into smaller functions so it is better readable :)

**Review by raphaelschwinger — COMMENTED** [review]
@XgamerTV Thanks for your work. Really good results!
Please add as a comment the source of where you find the code.
And please structure the code a bit into smaller functions so it is better readable :)

**Comment by The-Mats** [comment]
I pretty much just copied the code over from their github (I should definitely link that), removed stuff that's not needed and tried to get it to run which was mainly just feeding the correct params.  So I think it would be a waste of time to change/restructure the code much as I also only roughly know what's happening 🤔

File: configs/module/network/ssast.yaml (hunk 1 above)
Comment (The-Mats) [reply configs/module/network/ssast.yaml]: The parameters are now explained in ssast.yaml and the code itself

File: birdset/modules/models/ssast.py (hunk 2 above)
Comment (The-Mats) [reply birdset/modules/models/ssast.py]: Forward function is much simpler now

File: birdset/modules/models/ssast.py (hunk 3 above)
Comment (The-Mats) [reply birdset/modules/models/ssast.py]: I removed unneeded code and explained it more

File: birdset/modules/models/ssast.py (hunk 4 above)
Comment (The-Mats) [reply birdset/modules/models/ssast.py:139]: I don't think so my preprocessing function seems to work well

File: birdset/modules/models/ssast.py (hunk 5 above)
Comment (The-Mats) [reply birdset/modules/models/ssast.py:98]: No, that part is needed to init the backbone that is then used for finetuning

File: configs/module/network/ssast.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (raphaelschwinger) [inline configs/module/network/ssast.yaml]:
```suggestion
```

**Review by raphaelschwinger — APPROVED** [review]
@XgamerTV Thanks!

**Pull request merged**